README.md
<!-- README.md is generated from README.Rmd. Please edit that file -->
bcrm
====

Implements a wide variety of one and two-parameter Bayesian CRM designs. The program can run interactively, allowing the user to enter outcomes after each cohort has been recruited, or via simulation to assess operating characteristics.

Installation
------------

You can install from CRAN with:

``` r
install.packages("bcrm")
```

Or try the development version from GitHub with:

``` r
# install.packages("devtools")
devtools::install_github("mikesweeting/bcrm")
```

Example
-------

``` r
library(bcrm)

## Dose-escalation cancer trial example as described in Neuenschwander et al 2008.
## Pre-defined doses
dose<-c(1,2.5,5,10,15,20,25,30,40,50,75,100,150,200,250)
## Pre-specified probabilities of toxicity
## [dose levels 11-15 not specified in the paper, and are for illustration only]
p.tox0<-c(0.010,0.015,0.020,0.025,0.030,0.040,0.050,0.100,0.170,0.300,0.400,0.500,0.650
  ,0.800,0.900)
## Data from the first 5 cohorts of 18 patients
data<-data.frame(patient=1:18,dose=rep(c(1:4,7),c(3,4,5,4,2)),tox=rep(0:1,c(16,2)))
## Target toxicity level
target.tox<-0.30
## Random seed set to 12345 for this example
seed<-12345

## Simulate 10 replicate trials of size 36 (cohort size 3) using this design 
## with constraint (i.e. no dose-skipping) and starting at lowest dose
## True probabilities of toxicity are set to pre-specified probabilities (p.tox0) 
Power.LN.bcrm.sim<-bcrm(stop=list(nmax=36),p.tox0=p.tox0,dose=dose,ff="power"
  ,prior.alpha=list(3,0,1.34^2),target.tox=target.tox,constrain=TRUE
  ,sdose.calculate="median",pointest="mean",start=1,simulate=TRUE,nsims=10,truep=p.tox0, seed=seed)
#> Simulated trial: 10
print(Power.LN.bcrm.sim)
#> Operating characteristics based on  10  simulations: 
#>  
#>               
#> Sample size 36
#> 
#>                             Doses
#>                              No dose      1    2.5      5     10     15
#>   Experimentation proportion      NA 0.0833 0.0833 0.0833 0.0833 0.0833
#>   Recommendation proportion        0 0.0000 0.0000 0.0000 0.0000 0.0000
#>                             Doses
#>                                  20     25     30    40    50     75
#>   Experimentation proportion 0.0833 0.0833 0.0833 0.167 0.117 0.0417
#>   Recommendation proportion  0.0000 0.0000 0.0000 0.200 0.600 0.2000
#>                             Doses
#>                                  100 150 200 250
#>   Experimentation proportion 0.00833   0   0   0
#>   Recommendation proportion  0.00000   0   0   0
#> 
#>                             Probability of DLT
#>                              [0,0.2] (0.2,0.4] (0.4,0.6] (0.6,0.8] (0.8,1]
#>   Experimentation proportion   0.833     0.158   0.00833         0       0
#>   Recommendation proportion    0.200     0.800   0.00000         0       0
```


NEWS
Version 0.5.6 (15th July 2022)
-------------
* Missing "is.null(stop$safety)" from "bcrm" function included.
* Version numbers and dates updated.

Version 0.5.5 (31st May 2022)
-------------
* Bug fixed when in "posteriors.R": "dlnorm" at lines 171, 340 and 506 changed to rlnorm (thanks to Ling Zhang or University of Iowa for spotting).
* Version numbers and dates updated.

Version 0.5.4 (29th July 2019)
-------------
* Bug fixed when start not required (caused some runs of bcrm to stop).
* "target.tox" argument added to "stop.check" and "simFun"
* Version numbers and dates updated.

Version 0.5.3 (4th June 2019)
-------------
* Bug changing start dose when constrain = TRUE fixed.
* Version numbers and dates updated.

Version 0.5.2 (23rd May 2019)
-------------
* MCMC packages ("R2WinBUGS", "BRugs", "rjags") moved from "Import" to "Suggests" in Description file - many thanks to Prof. Brian Ripley for this request.
* "plot.bcrm" updated to draw points at DLT estimates used in trial
rather than at median only.
* Prior sampling under MCMC approaches moved directly into R rather than calling MCMC programs.
* Version numbers and dates updated.

Version 0.5.1 (3rd April 2019)
-------------
* Minor bug in "nextdose" fixed, to ensure correct dose
selected as starting dose if "start" not specified.
* README file updated/re-knitted.

Version 0.5.0 (1st April 2019)
-------------
* Restructuring of functions - many thanks to Harry Southworth for tidying up the package and reporting key bugs.
* "only.below" option added to "bcrm", such that if TRUE and loss not used, next dose chosen is that with risk of DLT closest to AND below the Target Toxicity Level.
* "threep3.start" option added to "bcrm" and "threep3", allowing specification of different start dose for 3+3 design with de-escalation permitted
* "threep3.esc.only" option added to "bcrm" and "threep3", such that if TRUE, runs the 3+3 design with no de-escalation permitted; if FALSE, de-escalation is permitted.


Version 0.4.9 (10th October 2018)
-------------
* Bug corrected in interactive version of bcrm to ensure updating of plots when option "plot = TRUE" is specified.
* Installation procedure would fail if an old version of rlang was used. The package "rlang" is now Imported at installation.
* In trial plots in interactive version of bcrm, previously only posterior median probabilities of DLT were shown as filled dots. The symbol for medians has been changed to black lines, and the user-specified estimate for the probability of DLT (plugin, mean) is shown as a filled point.


Version 0.4.8 (4th October 2018)
-------------
* bug corrected in interactive version of bcrm where over-ridden dose choices were not correctly recorded in output data.frame. Thanks to Graham Wheeler for spotting this.


Version 0.4.7 (28th January 2018)
-------------
* threep3 now works with any starting dose, not just the first. Thanks to Bernard North and Christina Yassouridis for spotting this bug.
* A safety stopping rule has now been implemented with stop(list(safety=p)) following the approach of Zohar and Chevret, Statistics in Medicine (2001)
* Updated bcrm help page. p.tox0 are the skeleton probabilities, not the standardised doses. Thanks to Eliezer Weill for spotting this.
* Using geom_bar to obtain percentage of subjects with DLTs since weight option removed from geom_hist
* threep3 example in manual corrected with named arguments


Version 0.4.6 (20th November 2015)
-------------
* Plotting functions updated to work with latest version of ggplot2 (>= 1.0.1)
* Argument "quietly" has been added to allow user to suppress simulation number output when running bcrm.
* print.bcrm.sim now returns a matrix of experimented dose proportions and a vector of recommended doses


Version 0.4.5 (23rd September 2015)
-------------
* The threep3 command also now returns all possible desigin in an object labelled `all.designs' for the user to interrogate further if required
* The threep3 command now also returns the average number of patients per dose level and the average number of DLTs per dose level
* Designs that use toxicity intervals now return the posterior probability of lying in each interval in the print.bcrm output
* seed argument added to allow reproducibility of results
* The help files for bcrm and getprior incorrectly specified the parameter `b' as the standard deviation of the lognormal prior. It is actually the variance. Many thanks to Elaidi Reza for spotting this error
* DESCRIPTION file details changed

Version 0.4.4 (18th March 2014)
-------------
* Maintainer contact details changed

Version 0.4.3 (10th September 2013)
-------------
* rjags can now be used instead of R2WinBUGS or BRugs
* Allows the 3+3 design to start at any dose level (a modified 3+3)
* Includes citation information for Journal of Statistical Software publication

Version 0.4.2 (19th March 2013)
-------------
* Corrects bug in simulation studies that do not use exact.sim computation, whereby an error is returned if more than 100 simulations are performed.
* Restricts bcrm and Posterior.BRugs examples to Windows and Linux Operating Systems where method="BRugs" is used. Hence Mac version should now compile on CRAN.

Version 0.4.1 (24th October 2012)
-------------
* New data argument allows specification of doses and toxicities of all previously recruited patients, and the sequential order in which they were entered into the trial. Arguments tox and notox in bcrm are now deprecated.
* The dose trajectory and observed toxicities from a single trial can now be plotted using the argument trajectory = TRUE inside plot.bcrm

Version 0.4 (18th September 2012)
-------------
* Simulations now allow the chosen Bayesian design to be compared alongside a traditional 3+3 algorithmic design.
* Corrects an error where simulation trajectories were not being plotted when specifying the argument trajectories = TRUE inside plot.bcrm.sim

Version 0.3 (1st May 2012)
-------------

R/bcrm-package.R
#' Bayesian Continual Reassessment Method for Phase I Dose-Escalation Trials
#' 
#' Implements a wide variety of Bayesian CRM designs. The program can run
#' interactively, allowing the user to enter outcomes after each cohort has
#' been recruited, or via simulation to assess operating characteristics.
#' 
#' \tabular{ll}{ Package: \tab bcrm\cr Type: \tab Package\cr Version: \tab
#' 0.5.5\cr Date: \tab 2022-05-31\cr License: \tab GPL (>= 2) \cr LazyLoad:
#' \tab yes\cr }
#' 
#' @name bcrm-package
#' @docType package
#' @author Michael Sweeting \email{michael.sweeting@@leicester.ac.uk}
#' 
#' Graham Wheeler \email{graham.wheeler@@imperial.ac.uk}
#' 
#' Maintainer: Graham Wheeler \email{graham.wheeler@@imperial.ac.uk}
#' @references Sweeting M., Mander A., Sabin T. \pkg{bcrm}: Bayesian Continual
#' Reassessment Method Designs for Phase I Dose-Finding Trials. \emph{Journal
#' of Statistical Software} (2013) 54: 1--26.
#' \doi{10.18637/jss.v054.i13}
NULL





#' Calculate posterior distribution of CRM model parameter(s)
#' 
#' Given the prior, functional form of the dose-toxicity model, and data, this
#' function returns the posterior distribution (via either MCMC samples, or
#' posterior summaries) of the CRM model parameter(s)
#' 
#' \code{Posterior.exact} produces posterior summary statistics of the CRM
#' model parameter(s), and probabilities of toxicity at the dose levels using
#' exact Bayesian computation (integration) of the posterior distribution. If
#' \code{Posterior.BRugs} or \code{Posterior.R2WinBUGS} is specified, then
#' posterior samples of the CRM model parameter(s) is returned by the function.
#' 
#' @aliases Posterior.exact Posterior.rjags Posterior.BRugs Posterior.R2WinBUGS
#' @param tox A vector of length \code{k} listing the number of patients who
#' have experienced the outcome (toxicity) at each dose level \code{1,...,k}.
#' If missing, then it is assumed that no data have thus far been collected.
#' @param notox A vector of length \code{k} listing the number of patients who
#' have not experienced the outcome (toxicity) at each dose level
#' \code{1,...,k}. If missing, then it is assumed that no data have thus far
#' been collected.
#' @param sdose A vector of length \code{k} listing the standardised doses used
#' in the CRM model.
#' @param ff A string indicating the functional form of the dose-response
#' curve. Options are \describe{ \item{ht}{ 1-parameter hyperbolic tangent}
#' \item{logit1}{ 1-parameter logistic} \item{power}{ 1-parameter power}
#' \item{logit2}{ 2-parameter logistic} }
#' @param prior.alpha A list of length 3 containing the distributional
#' information for the prior. The first element is a number from 1-4 specifying
#' the type of distribution. Options are \enumerate{ \item Gamma(a,b), where
#' a=shape, b=scale: mean=a*b, variance=a*b*b \item Uniform(a,b), where a=min,
#' b=max \item Lognormal(a,b), where a=mean on the log scale, b=standard
#' deviation on the log scale \item Bivariate Lognormal(a,b), where a=mean
#' vector on the log scale, b=Variance-covariance matrix on the log scale. This
#' prior should be used only in conjunction with a two-parameter logistic
#' model.  } The second and third elements of the list are the parameters a and
#' b, respectively.
#' @param burnin.itr Number of burn-in iterations (default 2000).
#' @param production.itr Number of production iterations (default 2000).
#' @param bugs.directory Directory that contains the WinBUGS executable if
#' \code{method="R2WinBUGS"}. Defaults to "C:/Program Files/WinBUGS14/".
#' @author Michael Sweeting \email{mjs212@@medschl.cam.ac.uk} (University of
#' Cambridge, UK)
#' @seealso \code{\link{bcrm}}
#' @references Sweeting M., Mander A., Sabin T. \pkg{bcrm}: Bayesian Continual
#' Reassessment Method Designs for Phase I Dose-Finding Trials. \emph{Journal
#' of Statistical Software} (2013) 54: 1--26.
#' \doi{10.18637/jss.v054.i13}
#' @examples
#' 
#' ## Dose-escalation cancer trial example as described in Neuenschwander et al 2008.
#' ## Pre-defined doses
#' dose<-c(1,2.5,5,10,15,20,25,30,40,50,75,100,150,200,250)
#' ## Pre-specified probabilities of toxicity
#' ## [dose levels 11-15 not specified in the paper, and are for illustration only]
#' p.tox0<-c(0.010,0.015,0.020,0.025,0.030,0.040,0.050,0.100,0.170,0.300,0.400,0.500,0.650
#'   ,0.800,0.900)
#' ## Data from the first 5 cohorts of 18 patients
#' tox<-c(0,0,0,0,0,0,2,0,0,0,0,0,0,0,0)
#' notox<-c(3,4,5,4,0,0,0,0,0,0,0,0,0,0,0)
#' ## Target toxicity level
#' target.tox<-0.30
#' ## Lognormal prior
#' prior.alpha<-list(3,0,1.34^2)
#' ## Power functional form
#' ff<-"power"
#' ## Standardised doses
#' sdose<-find.x(ff,p.tox0,alpha=1)
#' 
#' ## Posterior distribution of the model parameter using exact computation
#' post.exact<-Posterior.exact(tox,notox,sdose,ff,prior.alpha)
#' print(post.exact)
#' 
#' ## Posterior distribution of the model parameter using rjags
#' post.rjags<-Posterior.rjags(tox,notox,sdose,ff,prior.alpha
#'   ,burnin.itr=2000,production.itr=2000)
#' print(mean(post.rjags))
#' hist(post.rjags)
#' 
#' ## Posterior distribution of the model parameter using BRugs (Windows and i386 Linux only)
#' if(Sys.info()["sysname"] %in% c("Windows","Linux")){
#' 	post.BRugs<-Posterior.BRugs(tox,notox,sdose,ff,prior.alpha
#' 	  ,burnin.itr=2000,production.itr=2000)
#' 	print(mean(post.BRugs))
#' 	hist(post.BRugs)
#' 	}
#' 